### PR TITLE
perf: improve error handling

### PR DESCRIPTION
**Pull request by thisthat**
## This PR

While working with OF, I noticed that in case a flag is not found, the whole stack trace is attached to the exception.
I think this info is not necessary for this type of exception since it is hidden from the end user, and the SDK only requires the `ErrorCode` field. 

I quickly configured some JMH tests that use the InMemory FlagD provider to test the performance gains in evaluating a flag that doesn't exist. The Score value represents the amount of FF evaluations done in a single millisecond. 
In a single thread case, the performance gain is >60%. 

### Before

```
"Benchmark","Mode","Threads","Samples","Score","Score Error (99,9%)","Unit"
"booleanNotFound_01Thread","thrpt",1,10,"459,324808","8,254488","ops/ms"
"booleanNotFound_02Thread","thrpt",2,10,"791,244898","8,644711","ops/ms"
"booleanNotFound_05Thread","thrpt",5,10,"1465,736432","17,059916","ops/ms"
```

### After

```
"Benchmark","Mode","Threads","Samples","Score","Score Error (99,9%)","Unit"
"booleanNotFound_01Thread","thrpt",1,10,"765,474969","17,161956","ops/ms"
"booleanNotFound_02Thread","thrpt",2,10,"1159,854112","23,483354","ops/ms"
"booleanNotFound_05Thread","thrpt",5,10,"1828,460599","45,932954","ops/ms"
```

### CODE REVIEW

**Review by beeme1mr — APPROVED** [review]
Do we need the stack trace on any OpenFeatureError?

File: src/main/java/dev/openfeature/sdk/exceptions/FlagNotFoundError.java [hunk 1]
```diff
@@ -9,4 +9,9 @@
 public class FlagNotFoundError extends OpenFeatureError {
     private static final long serialVersionUID = 1L;
     @Getter private final ErrorCode errorCode = ErrorCode.FLAG_NOT_FOUND;
+
+    @Override
+    public Throwable fillInStackTrace() {
```
Comment (toddbaert) [inline src/main/java/dev/openfeature/sdk/exceptions/FlagNotFoundError.java:14]:
```suggestion
    public synchronized Throwable fillInStackTrace() {
```

Sonarcloud is recommending this: https://sonarcloud.io/project/issues?resolved=false&severities=BLOCKER%2CCRITICAL%2CMAJOR%2CMINOR&sinceLeakPeriod=true&types=BUG&pullRequest=739&id=open-feature_java-sdk&open=AYyHw5peA4eBhJGjCSRs

**Review by toddbaert — APPROVED** [review]
Agree with this. Just see [here](https://github.com/open-feature/java-sdk/pull/739/files#r1432917044).

File: src/main/java/dev/openfeature/sdk/exceptions/FlagNotFoundError.java (hunk 1 above)
Comment (thisthat) [reply src/main/java/dev/openfeature/sdk/exceptions/FlagNotFoundError.java:14]: oh true, the method in Throwable is sync 🤦

**Comment by toddbaert** [comment]
> Do we need the stack trace on any OpenFeatureError?

We have cases where consumers do things like:

```java
catch (IOException e) {
  throw new GeneralError("Something bad happend", e);
}
```

This is especially important for things like `GeneralError` where something unexpected has happened due to a logic error or system failure. I don't know how `fillInStackTrace` impacts this. I presume the "Caused by" normally seen in the stack-trace would be lost, and therefor I think in logs, etc, the underlying "cause" would be lost as well.

@thisthat

**Comment by thisthat** [comment]
@toddbaert the suggested `fillInStackTrace()` implementation prints only `dev.openfeature.sdk.exceptions.FlagNotFoundError` instead of the full list of methods calls from the main to the throw of the exception.
For end users, the stack trace is not relevant because they (hopefully) would never see any exception.
So I think we should look into the code of existing providers and how they use the exceptions.

**Comment by toddbaert** [comment]
> @toddbaert the suggested `fillInStackTrace()` implementation prints only `dev.openfeature.sdk.exceptions.FlagNotFoundError` instead of the full list of methods calls from the main to the throw of the exception. For end users, the stack trace is not relevant because they (hopefully) would never see any exception. So I think we should look into the code of existing providers and how they use the exceptions.

@thisthat  Yes, I understand that. I was making a point relative to @beeme1mr 's question:

> Do we need the stack trace on any OpenFeatureError?

My concern is that if we do this for other exceptions which are more complicated than `FlagNotFound`, we lose the causal information.

**Pull request merged**